---
title: "Genotyping QC"
output: html_notebook
---

- See the [wiki](http://wiki.psycm.cf.ac.uk/wiki/index.php/GWAS_Quality_control_(QC)_pipeline) for QC steps

- add sex for samples with missing metadata
- 12994 genotyped twice, need to manually filter out duplicate rows after going with VCFindex

```{r setup}
library(tidyverse)
library(RColorBrewer)
source("..//Shiny/GENEX-FB2/FormatGGplot.R")
VCFindex <- read_delim("~/BTSync/FetalRNAseq/LabNotes/VCFindex.txt", 
     "\t", escape_double = FALSE, col_names = c('Sample', 'Index', 'VCF'), 
     trim_ws = TRUE)
SampleInfo <- read_delim("~/BTSync/FetalRNAseq/Github/GENEX-FB1/Data/SampleInfo.txt", 
     "\t", escape_double = FALSE, col_types = cols(Sample = col_character()), 
     trim_ws = TRUE)

key <- read_delim("/Users/heo3/BTSync/FetalRNAseq/Genome-wide\ genotyping/key.txt", 
     "\t", escape_double = FALSE, col_names = c('Sample', 'IID'), 
     trim_ws = TRUE)

sample_info <- read_delim("~/BTSync/FetalRNAseq/LabNotes/sample_info.txt", 
     "\t", escape_double = FALSE, col_types = cols(BrainBankID = col_character(), 
         PCW = col_double()), trim_ws = TRUE)

all_sample_info <- sample_info %>% mutate(PCW=floor(PCW)) %>% 
  dplyr::select(Sample=BrainBankID, Sex, PCW) %>% 
  anti_join(SampleInfo, by="Sample") %>% 
  bind_rows(mutate(SampleInfo, PCW=as.numeric(PCW))) %>% 
  full_join(key) %>% 
  full_join(VCFindex) %>% 
  mutate(FID=paste0(VCF, Index)) %>%
  filter(!(IID == '200720170085_R04C02' & Index == 32)) %>%
  filter(!(IID == '200720170117_R10C01' & Index == 19)) %>%
  dplyr::select(Sample, FID, IID, VCF, everything()) %>%
  mutate(Sex=ifelse(IID %in% c('200720170117_R05C01', '200720170117_R12C01', '200720170085_R10C02'), 'Male',
                    ifelse(IID %in% c('200720170117_R06C01', '200720170117_R11C01', '200720170117_R11C02'), 
                           'Female', Sex)))
group_by(all_sample_info, Sample) %>% summarise(n=n()) %>% filter(n> 1)
write_tsv(all_sample_info, "~/BTSync/FetalRNAseq/Github/GENEX-FB2/Data/SampleInfo.txt")

Omni25_regions_With_Code <- read_delim("../Genotyping/1KGP_Heath/Omni25.regions_With_Code.txt", 
     "\t", escape_double = FALSE, col_names = FALSE, 
     trim_ws = TRUE) %>% full_join(read_delim("../Genotyping/1KGP_Heath/Omni25.regions_With_Gpop_Code.txt", 
     "\t", escape_double = FALSE, col_names = FALSE, 
     trim_ws = TRUE), by='X1'
) %>%
  dplyr::select(IID=X1, POP=X2.x, SuperPop=X2.y, PopNum=X3.x)
all_sample_info <- mutate(all_sample_info, POP=VCF, SuperPop=VCF) %>% full_join(Omni25_regions_With_Code)

# use final expression matrix to get list of samples
# 12994 was genotyped twice. We used the data from the first sample (FB_new19/200720170085_R04C02)
included_samples <- all_sample_info %>% semi_join(tibble(Sample=colnames(counts_vst)[-1])) %>%
  group_by(Sample) %>% dplyr::slice(1) %>% ungroup()
```

# Merge Plink files
- This produces files where there are two-three samples with each FID (two from the previously merged file plus one from the new file)
- I will have to be careful about doing analyses that use this information
- I also need the FIDs to update the Sex info.
- The safest thing to do is probably to update the FIDs so they are unique for each sample
- The easiest way that I know of to extract FIDs is to run a heterozygosity analysis

```{bash plinkMerge}

cd ../Genotyping

# warnings are printed to log, so they don't need to be echoed here
# convert to binary
~/bin/plink --file PLINK_040517_1054/Nick_Bray_OmniExpress_230317 --make-bed --out PLINK_040517_1054/FB_new > /dev/null 2>&1

# remove blank samples
echo 24 > excluded_fam.txt
seq 45 48 >> excluded_fam.txt
~/bin/plink --bfile PLINK_040517_1054/FB_new --remove-fam excluded_fam.txt --make-bed --out PLINK_040517_1054/FB_new_excl_blank > /dev/null 2>&1

# merge
mkdir Merged
#echo FB_Merged >all_my_files.txt
~/bin/plink --bfile PLINK_040517_1054/FB_new_excl_blank --bmerge Set1/FB_Merged --make-bed --out Merged/FB_all > /dev/null 2>&1

~/bin/plink --bfile Merged/FB_all --het --out Merged/FB_all_het  > /dev/null 2>&1
```

```{r getFIDs}
HET = read.table("../Genotyping/Merged/FB_all_het.het", h=T, as.is=T)

newFID <- all_sample_info %>% 
  mutate(newIID=IID) %>% 
  dplyr::select(IID, newFID=FID, newIID) %>% 
  right_join(HET) %>% 
  dplyr::select(FID, IID, newFID, newIID)
write_tsv(newFID, "../Genotyping/Merged/newFID.txt", col_names=FALSE)
```

``` {bash plinkFID}
# warnings are printed to log, so they don't need to be echoed here
cd ../Genotyping
~/bin/plink --bfile Merged/FB_all --update-ids Merged/newFID.txt --make-bed --out Merged/FB_all_newFID > /dev/null 2>&1
```

# Add Sex info to PLINK file
```{bash plinkSex}
cd ../Genotyping
# add Sex
cat ../Data/SampleInfo.txt | cut -f 2,3,5 | perl -pe 's/Male/1/; s/Female/2/; s/NA/0/' > Merged/gender.txt  
~/bin/plink --bfile Merged/FB_all_newFID --update-sex Merged/gender.txt --make-bed --out  Merged/FB_all_with_sex > /dev/null 2>&1

~/bin/plink --bfile Merged/FB_all_with_sex --check-sex --out Merged/FB_all_with_sex_check > /dev/null 2>&1
```
- 4 samples have intermediate sex genotypes, one of which (18655) is included in the sex diffs study
    - This sample clusters with other Females in PCA
    -  I've already tried excluding it because of the sequencing depth and it made very little difference
- 1 sample (17118) is coded as Male but with a female genotype
``` {r SexCheck}
SexCheck <- read.table("../Genotyping/Merged/FB_all_with_sex_check.sexcheck", h=T, as.is=T)
full_join(SexCheck, all_sample_info) %>% View()
right_join(SexCheck, included_samples) %>% View()
```

```{bash plinkMerge1kg}
cd ../Genotyping

# warnings are printed to log, so they don't need to be echoed here
~/bin/plink --bfile Merged/FB_all_with_sex --bmerge 1KGP_Heath/omni25.2141_1KG_HeathSNPs --make-bed --out Merged/FB_all_1kg > /dev/null 2>&1

~/bin/plink --bfile 1KGP_Heath/omni25.2141_1KG_HeathSNPs --flip Merged/FB_all_1kg-merge.missnp --make-bed --out 1KGP_Heath/omni25.2141_1KG_HeathSNPs_flip  > /dev/null 2>&1

~/bin/plink --bfile Merged/FB_all_with_sex --bmerge 1KGP_Heath/omni25.2141_1KG_HeathSNPs_flip --make-bed --out Merged/FB_all_1kg2  > /dev/null 2>&1

~/bin/plink --bfile 1KGP_Heath/omni25.2141_1KG_HeathSNPs_flip --exclude Merged/FB_all_1kg2-merge.missnp --make-bed --out 1KGP_Heath/omni25.2141_1KG_HeathSNPs_flip_excl  > /dev/null 2>&1

~/bin/plink --bfile Merged/FB_all_with_sex --bmerge 1KGP_Heath/omni25.2141_1KG_HeathSNPs_flip_excl --make-bed --out Merged/FB_all_1kg3  > /dev/null 2>&1

```


# Investigate missingness rates

```{bash plinkMiss}

cd ../Genotyping

~/bin/plink --bfile Merged/FB_all_1kg3 --missing --out Merged/FB_all_1kg3_miss  > /dev/null 2>&1
```
- A225 has a call rate of 83.5%, 17751 has a call rate of 89.8% and 18983 has a call rate of 92.8%
- 8 samples have call rates between 95% and 99%
- None have call rates >99%
- When looking just at new samples, almost all were >99%, so this must be due to merging files with different SNPs

- Most of our samples have less missing loci than any of the 1kg samples. This is because Alex gave me a file with only my SNPs in it and the reference panel is clearly not typed for all of our SNPs

```{r IMISS}
IMISS = read.table("../Genotyping/Merged/FB_all_1kg3_miss.imiss", h=T, as.is=T)

IMISS <- IMISS %>%
  arrange(desc(F_MISS)) %>%
  mutate(rank=row_number()) %>% 
  left_join(all_sample_info, by="IID")

ggplot(filter(IMISS, is.na(VCF)), aes(x=rank, y=1-F_MISS)) + 
  geom_point(aes(colour=SuperPop), alpha=.1) +
  geom_point(size=.5, data=filter(IMISS, !is.na(VCF))) +
  geom_text(aes(label=Sample), colour='black', hjust=0, nudge_x=20, data=filter(IMISS, !is.na(VCF) & F_MISS> 0.04)) +
  geom_hline(yintercept=.98) +
  main_theme() +
  theme(legend.position=c(.9,.5)) +
  scale_colour_brewer(type = "qual", palette = 6) +
  scale_y_continuous(limits=c(NA, 1)) +
  guides(colour = guide_legend(override.aes = list(alpha = 1)))


filter(IMISS, F_MISS > .02) %>% left_join(all_sample_info) %>% arrange(desc(F_MISS))
filter(IMISS, F_MISS > .02) %>% inner_join(included_samples) %>% arrange(desc(F_MISS)) %>% View()
```
- 53% of SNPs are complete
- 31% have one missing call
- 96% have <2% missing calls

- These numbers are a lot lower after adding 1kg data (5%, 10% and 91.5%)
```{r LMISS}
LMISS = read.table("../Genotyping/Merged/FB_all_1kg3_miss.lmiss", h=T, as.is=T)
LMISS %>%
  arrange(desc(F_MISS)) %>%
  mutate(rank=row_number()) %>% 
  ggplot(aes(x=rank, y=1-F_MISS)) + 
  geom_point(alpha=.1, size=.5) + 
  geom_hline(yintercept =0.98 ) +
  main_theme() +
  theme(legend.position=c(.9,.8)) +
  scale_colour_brewer(type = "qual", palette = 6)


group_by(LMISS, N_MISS) %>% summarise(perc=(n()/nrow(LMISS))*100)
filter(LMISS, F_MISS<0.02) %>% dim()/dim(LMISS)
```


# Minor Allele Frequency
```{bash plinkMAF}
cd ../Genotyping
~/bin/plink --bfile Merged/FB_all_1kg3 --freq --out Merged/FB_all_1kg_maf > /dev/null 2>&1
```

```{r MAF}
FREQ = read.table("../Genotyping/Merged/FB_all_1kg_maf.frq", h=T, as.is=T)

ggplot(FREQ, aes(x=MAF)) + 
  geom_freqpoly() +
  main_theme() +
  theme(legend.position=c(.9,.8)) +
  scale_colour_brewer(type = "qual", palette = 6)
```

# Heterozygosity

```{bash plink_heterozygosity}
cd ../Genotyping
~/bin/plink --bfile Merged/FB_all_1kg3 --het --out Merged/FB_all_1kg_het > /dev/null 2>&1
```

- A225 clearly has issues. This was also one of the samples that had intermediate sex

```{r Heterozygosity}
HET = read.table("../Genotyping/Merged/FB_all_1kg_het.het", h=T, as.is=T)
HET <- HET %>% mutate(H=(N.NM. - O.HOM.) / N.NM.) %>% full_join(all_sample_info, by='IID')
ggplot(HET, aes(x=H)) +
  geom_histogram() +
  geom_text(aes(label=Sample), y=2, data=filter(HET, H>0.4), angle=-45, vjust=0, hjust=1) +
  main_theme() +
  theme(legend.position=c(.9,.8)) +
  scale_colour_brewer(type = "qual", palette = 6)+
  facet_grid(VCF ~ .)


ggplot(HET, aes(x=F)) + 
  geom_histogram() +
  geom_text(aes(label=Sample), y=2, data=filter(HET, F< -0.3), angle=45, vjust=0, hjust=0) +
  main_theme() +
  theme(legend.position=c(.9,.8)) +
  scale_colour_brewer(type = "qual", palette = 6) +
  facet_grid(VCF ~ .)

```

# Hardy-Weinberg Equilibrium
- very few loci violating HWE. This might be down to small sample size

- When the 1kg data are included, over half of loci violate HWE. This probably isn't too surprising given population stratification

```{bash plink_hwe}
cd ../Genotyping/
~/bin/plink --bfile Merged/FB_all_1kg3 --hardy --out Merged/FB_all_1kg_hwe  > /dev/null 2>&1
```
- 6.6% of loci fail Hardy-Weinberg with an FDR<0.1
```{r HWE}
HWE = read.table("../Genotyping/Merged/FB_all_1kg_hwe.hwe", h=T, as.is=T)
ggplot(HWE, aes(x=log10(P))) + 
  geom_histogram() +
  main_theme() +
  theme(legend.position=c(.9,.8)) +
  scale_colour_brewer(type = "qual", palette = 6)

HWE$padj <- p.adjust(HWE$P, method='BH')
filter(HWE, padj<.1) %>% dim()/dim(HWE)
```

# PCA

```{bash plinkPCA}
cd ../Genotyping

~/bin/plink --bfile Merged/FB_all_1kg3 --indep-pairwise 250 5 0.2 --out Merged/FB_all_1kg_indep > /dev/null 2>&1

~/bin/plink --bfile Merged/FB_all_1kg3 --pca --extract Merged/FB_all_1kg_indep.prune.in --out Merged/FB_all_1kg_pca > /dev/null 2>&1
```

- 18655 also looks like it belongs to a different population
``` {r PCA}
PCA = read_delim("../Genotyping/Merged/FB_all_1kg_pca.eigenvec", ' ', col_names=c('FID', 'IID', paste0('PC', 1:20))) 

PCA <- full_join(all_sample_info, PCA, by=c('IID'))


EUR_mean <- filter(PCA, SuperPop=='EUR') %>% 
  dplyr::select(starts_with('PC'), -PCW) %>% 
  summarise_all(mean, na.rm=TRUE)

EUR_sd <- filter(PCA, SuperPop=='EUR') %>% 
  dplyr::select(starts_with('PC'), -PCW) %>% 
  summarise_all(sd, na.rm=TRUE)

# not sure how many PCs to use for this
EUR_samples <- semi_join(PCA, SampleInfoCombined) %>% 
                       filter(PC1 > EUR_mean[[1]]-2*EUR_sd[[1]] & PC1 < EUR_mean[[1]]+2*EUR_sd[[1]] &
                       PC2 > EUR_mean[[2]]-2*EUR_sd[[2]] & PC2 < EUR_mean[[2]]+2*EUR_sd[[2]])# &
                       #PC3 > EUR_mean[[3]]-2*EUR_sd[[3]] & PC3 < EUR_mean[[3]]+2*EUR_sd[[3]] &
                       #PC4 > EUR_mean[[4]]-2*EUR_sd[[4]] & PC4 < EUR_mean[[4]]+2*EUR_sd[[4]] &
                       #PC5 > EUR_mean[[5]]-2*EUR_sd[[5]] & PC5 < EUR_mean[[5]]+2*EUR_sd[[5]])

ggplot(filter(PCA, is.na(VCF)), aes(x=PC1, y=PC2, colour=SuperPop)) +
  geom_point(alpha=.2) +
  geom_point(colour='gray', size=.5, data=semi_join(PCA, SampleInfoCombined)) +
  geom_point(colour='black', size=.5, data=EUR_samples) +
  geom_text(aes(label=Sample, x=PC1+.001), colour='black', hjust=0, data=filter(PCA, Sample=='A225')) +
  xlab('PCA1') +
  ylab('PCA2') +
  main_theme() +
  theme(legend.position=c(.9,.8)) +
  scale_colour_brewer(type = "qual", palette = 6) +
  guides(colour = guide_legend(override.aes = list(alpha = 1)))

filter(PCA, SuperPop=='EUR') %>% ggplot(aes(x=PC1, y=PC2, colour=POP)) +
  geom_point(alpha=.2) +
  geom_point(colour='grey', size=.5, data=filter(PCA, !is.na(VCF))) +
  geom_point(colour='black', size=.5, data=EUR_samples) +
  geom_text(aes(label=Sample, x=PC1+.001), colour='black', hjust=0, data=filter(PCA, Sample=='A225')) +
  xlab('PCA1') +
  ylab('PCA2') +
  main_theme() +
  theme(legend.position=c(.9,.8)) +
  scale_colour_brewer(type = "qual", palette = 6) +
  guides(colour = guide_legend(override.aes = list(alpha = 1))) +
  scale_x_continuous(limits=c(-.01, -0.0075)) +
  scale_y_continuous(limits=c(0,.03))


SampleInfoCombined

ggsave("~/Desktop/PCA.png")  
```

```{bash}
cd ../Genotyping
echo FB_merged48 > Merged/exclude_A225.txt
~/bin/plink --bfile Merged/FB_all_1kg3 --remove-fam Merged/exclude_A225.txt --make-bed --out Merged/FB_new_1kg_exclA225 > /dev/null 2>&1

```